### PR TITLE
TF-3298 Fix blink when refresh email list

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#3298 

## Root cause

According to the response returned from the server, the number of elements in `Email/get` and `Email/query` are different. Because some elements are not found in the `notFound` list. This leads to the `sortingByOrderOfIdList` function returning the list from `Email/get`.

```json
{
    "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
    "methodResponses": [
        [
            "Email/query",
            {
                "accountId": "cce2ec1540fa8f369c1f75f2151e8c7a2364a158881c0fed63c068aae0015901",
                "queryState": "b09ff786",
                "canCalculateChanges": false,
                "ids": [
                    "2c503360-a8e7-11ef-8f35-2b6b7de002f9",
                    "14d6a350-a8e6-11ef-8f35-2b6b7de002f9",
                    "3f42c470-a8d7-11ef-9ab4-c903a7f1e8a3",
                    "f8f0aa80-a8ba-11ef-9340-0d8f5a8bd4f9",
                    "2c21f0e0-a8ba-11ef-9340-0d8f5a8bd4f9",
                    "2492b9c0-a88f-11ef-9340-0d8f5a8bd4f9",
                    "1f069710-a88f-11ef-9ab4-c903a7f1e8a3",
                    "1a14b890-a88f-11ef-9340-0d8f5a8bd4f9",
                    "441861f0-da0e-11ee-9863-a368905d3405",
                    "f5890ba0-d81b-11ee-a7f2-0302da82fa0d",
                    "a48f37b0-d81b-11ee-9863-a368905d3405",
                    "858b4ae0-c9eb-11ee-b31c-939afe0b6132",
                    "6603a830-c9ea-11ee-bfc8-8baa82bb865e",
                    "365c4150-c9ea-11ee-bfc8-8baa82bb865e",
                    "642a3370-c9cd-11ee-8e22-3b54566f9f90",
                    "541d5ad0-c9a9-11ee-9073-63121fc97643",
                    "a4dbc8b0-c983-11ee-a89a-a70d425dc805",
                    "86ded7f0-c972-11ee-a89a-a70d425dc805",
                    "c93de160-c970-11ee-9073-63121fc97643",
                    "0d4d3d70-c970-11ee-9073-63121fc97643"
                ],
                "position": 0
            },
            "c0"
        ],
        [
            "Email/get",
            {
                "accountId": "cce2ec1540fa8f369c1f75f2151e8c7a2364a158881c0fed63c068aae0015901",
                "state": "bbe31d30-a8e7-11ef-9ab4-c903a7f1e8a3",
                "list": [
                    {
                        "id": "1a14b890-a88f-11ef-9340-0d8f5a8bd4f9",
                        "blobId": "1a14b890-a88f-11ef-9340-0d8f5a8bd4f9",
                        "keywords": {},
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 1407,
                        "receivedAt": "2024-11-22T05:03:24Z",
                        "to": [
                            {
                                "name": "sonde-supervision-preprod.linagora@preprod-avocat.fr",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "signature 2",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "replyTo": [
                            {
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "subject": "26",
                        "sentAt": "2024-11-22T05:03:15Z",
                        "hasAttachment": false,
                        "preview": "--  Hello Signature 2",
                        "header:X-MEETING-UID:asText": null
                    },
                    {
                        "id": "1f069710-a88f-11ef-9ab4-c903a7f1e8a3",
                        "blobId": "1f069710-a88f-11ef-9ab4-c903a7f1e8a3",
                        "keywords": {},
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 1407,
                        "receivedAt": "2024-11-22T05:03:32Z",
                        "to": [
                            {
                                "name": "sonde-supervision-preprod.linagora@preprod-avocat.fr",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "signature 2",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "replyTo": [
                            {
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "subject": "27",
                        "sentAt": "2024-11-22T05:03:24Z",
                        "hasAttachment": false,
                        "preview": "--  Hello Signature 2",
                        "header:X-MEETING-UID:asText": null
                    },
                    {
                        "id": "2c21f0e0-a8ba-11ef-9340-0d8f5a8bd4f9",
                        "blobId": "2c21f0e0-a8ba-11ef-9340-0d8f5a8bd4f9",
                        "keywords": {
                            "$seen": true
                        },
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 4516,
                        "receivedAt": "2024-11-22T10:11:42Z",
                        "to": [
                            {
                                "name": "Test preprod avocat.fr",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "Patrick PEREIRA",
                                "email": "ppereira@linagora.com"
                            }
                        ],
                        "replyTo": [
                            {
                                "email": "ppereira@linagora.com"
                            }
                        ],
                        "subject": "Hello",
                        "sentAt": "2024-11-22T10:11:24Z",
                        "hasAttachment": false,
                        "preview": "Coucou",
                        "header:X-MEETING-UID:asText": null
                    },
                    {
                        "id": "14d6a350-a8e6-11ef-8f35-2b6b7de002f9",
                        "blobId": "14d6a350-a8e6-11ef-8f35-2b6b7de002f9",
                        "keywords": {
                            "$seen": true
                        },
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 4891,
                        "receivedAt": "2024-11-22T15:26:01Z",
                        "to": [
                            {
                                "name": "sonde-supervision-preprod.linagora@preprod-avocat.fr",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "Benoit TELLIER",
                                "email": "btellier@linagora.com"
                            }
                        ],
                        "replyTo": [
                            {
                                "email": "btellier@linagora.com"
                            }
                        ],
                        "subject": "erewrfwe",
                        "sentAt": "2024-11-22T15:25:41Z",
                        "hasAttachment": false,
                        "preview": "qwefqwef--  Best regards, Benoit TELLIER General manager of Linagora VIETNAM. Product owner for Team-Mail product. Chairman of the Apache James project. Mail: btellier@linagora.com Tel: (0033) 6 77 26 04 58 (WhatsApp, Signal)",
                        "header:X-MEETING-UID:asText": null
                    },
                    {
                        "id": "2c503360-a8e7-11ef-8f35-2b6b7de002f9",
                        "blobId": "2c503360-a8e7-11ef-8f35-2b6b7de002f9",
                        "keywords": {
                            "$seen": true
                        },
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 6461,
                        "receivedAt": "2024-11-22T15:33:50Z",
                        "to": [
                            {
                                "name": "sonde-supervision-preprod.linagora@preprod-avocat.fr",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "Benoit TELLIER",
                                "email": "btellier@linagora.com"
                            }
                        ],
                        "replyTo": [
                            {
                                "email": "btellier@linagora.com"
                            }
                        ],
                        "subject": "Re: erewrfwe",
                        "sentAt": "2024-11-22T15:33:31Z",
                        "hasAttachment": false,
                        "preview": "--  Best regards, Benoit TELLIER General manager of Linagora VIETNAM. Product owner for Team-Mail product. Chairman of the Apache James project. Mail: btellier@linagora.com Tel: (0033) 6 77 26 04 58 (WhatsApp, Signal) On Nov 22, 2024 4:25 PM, from Benoit T",
                        "header:X-MEETING-UID:asText": null
                    },
                    {
                        "id": "f8f0aa80-a8ba-11ef-9340-0d8f5a8bd4f9",
                        "blobId": "f8f0aa80-a8ba-11ef-9340-0d8f5a8bd4f9",
                        "keywords": {
                            "$seen": true
                        },
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 400078,
                        "receivedAt": "2024-11-22T10:17:26Z",
                        "to": [
                            {
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "Patrick PEREIRA",
                                "email": "ppereira@linagora.com"
                            }
                        ],
                        "replyTo": [
                            {
                                "email": "ppereira@linagora.com"
                            }
                        ],
                        "subject": "Re : test S3 -> ppereira",
                        "sentAt": "2024-11-22T10:17:08Z",
                        "hasAttachment": true,
                        "preview": "Le nov. 22, 2024 11:16 AM, de Signature 2 --  Hello Signature 2",
                        "header:X-MEETING-UID:asText": null
                    },
                    {
                        "id": "3f42c470-a8d7-11ef-9ab4-c903a7f1e8a3",
                        "blobId": "3f42c470-a8d7-11ef-9ab4-c903a7f1e8a3",
                        "keywords": {
                            "$seen": true
                        },
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 132052,
                        "receivedAt": "2024-11-22T13:39:50Z",
                        "to": [
                            {
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "Twake Drive",
                                "email": "noreply@preprod-avocat.fr"
                            }
                        ],
                        "subject": "Alerte Antivirus dans votre Twake Drive",
                        "sentAt": "2024-11-22T13:39:41Z",
                        "hasAttachment": false,
                        "preview": "Important: Alerte Antivirus dans votre Twake Drive L'analyse antivirus d'un fichier dans votre Twake Drive a signalé un problème. Fichier: eicar_com.zip Problème: Contenu malveillant détecté ⚠️ Voir sur Twake Drive Retrouvez nous sur les réseaux [instagram",
                        "header:X-MEETING-UID:asText": null
                    },
                    {
                        "id": "2492b9c0-a88f-11ef-9340-0d8f5a8bd4f9",
                        "blobId": "2492b9c0-a88f-11ef-9340-0d8f5a8bd4f9",
                        "keywords": {
                            "$seen": true
                        },
                        "mailboxIds": {
                            "4b028e10-c5c2-11ee-82c9-61411f1e5bf8": true
                        },
                        "size": 1407,
                        "receivedAt": "2024-11-22T05:03:41Z",
                        "to": [
                            {
                                "name": "sonde-supervision-preprod.linagora@preprod-avocat.fr",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "from": [
                            {
                                "name": "signature 2",
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "replyTo": [
                            {
                                "email": "sonde-supervision-preprod.linagora@preprod-avocat.fr"
                            }
                        ],
                        "subject": "28",
                        "sentAt": "2024-11-22T05:03:33Z",
                        "hasAttachment": false,
                        "preview": "--  Hello Signature 2",
                        "header:X-MEETING-UID:asText": null
                    }
                ],
                "notFound": [
                    "0d4d3d70-c970-11ee-9073-63121fc97643",
                    "6603a830-c9ea-11ee-bfc8-8baa82bb865e",
                    "642a3370-c9cd-11ee-8e22-3b54566f9f90",
                    "a48f37b0-d81b-11ee-9863-a368905d3405",
                    "365c4150-c9ea-11ee-bfc8-8baa82bb865e",
                    "c93de160-c970-11ee-9073-63121fc97643",
                    "541d5ad0-c9a9-11ee-9073-63121fc97643",
                    "441861f0-da0e-11ee-9863-a368905d3405",
                    "86ded7f0-c972-11ee-a89a-a70d425dc805",
                    "858b4ae0-c9eb-11ee-b31c-939afe0b6132",
                    "f5890ba0-d81b-11ee-a7f2-0302da82fa0d",
                    "a4dbc8b0-c983-11ee-a89a-a70d425dc805"
                ]
            },
            "c1"
        ]
    ]
}

```

cause-sort

## Solution

- Fix the `sortEmailsById` function to always return a list of emails from `Email/get` sorted by `Email/query` regardless of the number of elements in `Email/query`.

```dart
  List<Email> sortEmailsById(List<Id> referenceIds) {
    final indexMap = {
      for (var i = 0; i < referenceIds.length; i++)
        referenceIds[i]: i
    };

    sort((email1, email2) {
      final indexEmail1 = indexMap[email1.id!.id] ?? double.maxFinite;
      final indexEmail2 = indexMap[email2.id!.id] ?? double.maxFinite;

      return indexEmail1.compareTo(indexEmail2);
    });

    return this;
  }
```

## Resolved

- Passed unit test for `sortEmailsById` function

### CODE REVIEW

File: lib/features/thread/data/extensions/list_email_extension.dart [hunk 1]
```diff
@@ -17,23 +17,17 @@ extension ListEmailExtension on List<Email> {
...
-      final id2 = email2.id?.id;
-
-      if (id1 == null || id2 == null) {
-        return 0;
-      }
-
-      final index1 = ids.indexWhere((id) => id == id1);
-      final index2 = ids.indexWhere((id) => id == id2);
+      final indexEmail1 = indexMap[email1.id!.id] ?? double.maxFinite;
+      final indexEmail2 = indexMap[email2.id!.id] ?? double.maxFinite;
```
Comment (hoangdat) [inline lib/features/thread/data/extensions/list_email_extension.dart:28]: `!` is not safe here. How about `?`?

**Comment by chibenwa** [comment]
Ok, we encountered that because the probe see a very high volume of traffic.

Thus the email index is way more subject to being screwed up.

Regular users should not be affected.

Thanks for the quick reaction and for having a look @hoangdat @dab246 !

**Review by hoangdat — DISMISSED** [review]

**Review by tddang-linagora — DISMISSED** [review]

File: lib/features/thread/data/extensions/list_email_extension.dart (hunk 1 above)
Comment (dab246) [reply lib/features/thread/data/extensions/list_email_extension.dart:28]: It is safe, as per jmap documentation. Email always has id, so it is always non-null.

Comment (hoangdat) [reply lib/features/thread/data/extensions/list_email_extension.dart:28]: But in client side it is nullable, it also depends in get properties, so it should be safer in our side

Comment (dab246) [reply lib/features/thread/data/extensions/list_email_extension.dart:28]: So we should change them all to a non-null property. That's the right thing to do. We shouldn't care too much about the cases when it's optional.

Comment (dab246) [reply lib/features/thread/data/extensions/list_email_extension.dart]: updated

**Review by hoangdat — APPROVED** [review]
Please rebase

**Review by tddang-linagora — APPROVED** [review]

**Pull request merged**